internal/exec/exec_test.go

```diff
@@ -0,0 +1,257 @@
+package exec
+
+import (
+	"encoding/binary"
+	"os"
+	"testing"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"github.com/Microsoft/hcsshim/osversion"
+	_ "github.com/Microsoft/hcsshim/test/functional/manifest"
+	"golang.org/x/sys/windows"
+)
+
+// Tests that the standard os/exec.Cmd functionality works the same.
+func TestExecProcess(t *testing.T) {
+	cmd := Command("ping", "127.0.0.1")
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+// Tests that the new procattrlist functionality works as expected.
+func TestExecProcessAttrList(t *testing.T) {
+	procAttrList, err := winapi.CreateProcThreadAttrList(1)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	// This is useless but it tests all the win32 calls working. A better use for this test would be creating
+	// this process to be a child of a different process other than the one launching it but we can't gurantee
+	// any other processes running/get their pids.
+	pHandle, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, false, uint32(os.Getpid()))
+	if err != nil {
+		t.Fatalf("failed to open process: %s", err)
+	}
+	uintPtrHandle := uintptr(pHandle)
+	if err := winapi.UpdateProcThreadAttribute(
+		procAttrList,
+		0,
+		winapi.PROC_THREAD_ATTRIBUTE_PARENT_PROCESS,
+		&uintPtrHandle,
+		unsafe.Sizeof(pHandle),
+		nil,
+		nil,
+	); err != nil {
+		t.Fatalf("UpdateProcThreadAttribute failed: %s", err)
+	}
+
+	cmd := Command("ping", "127.0.0.1")
+	cmd.SysProcAttr = &SysProcAttr{
+		ProcThreadAttrList: procAttrList,
+	}
+
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+func TestJobObject(t *testing.T) {
+	procAttrList, err := winapi.CreateProcThreadAttrList(1)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	job, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr("test"))
+	if err != nil {
+		t.Fatalf("failed to create job object")
+	}
+	uintPtrHandle := uintptr(job)
+	if err := winapi.UpdateProcThreadAttribute(
+		procAttrList,
+		0,
+		winapi.PROC_THREAD_ATTRIBUTE_JOB_LIST,
+		&uintPtrHandle,
+		unsafe.Sizeof(job),
+		nil,
+		nil,
+	); err != nil {
+		t.Fatalf("UpdateProcThreadAttribute failed: %s", err)
+	}
+
+	cmd := Command("ping", "127.0.0.1")
+	cmd.SysProcAttr = &SysProcAttr{
+		ProcThreadAttrList: procAttrList,
+	}
+
+	if err := cmd.Run(); err != nil {
+		t.Fatalf("failed to run process: %s", err)
+	}
+}
+
+func TestPseudoConsole(t *testing.T) {
+	// Skip if not on RS5. This is when the pseudoconsole API was introduced
+	if osversion.Get().Build < osversion.RS5 {
+		t.Skip()
+	}
+
+	cmd := Command("ping", "127.0.0.1")
+
+	var b [4]byte
+	binary.LittleEndian.PutUint16(b[0:], 80)
+	binary.LittleEndian.PutUint16(b[2:], 40)
+	coord := uintptr(binary.LittleEndian.Uint32(b[:]))
```

Comment:
I don't really understand what this is doing. Where do the values for 80 and 40 come from?

Reply:
height/width of the pseudo console. They could be anything honestly.

Reply:
CreatePseudoConsole expects a COORD structure (https://docs.microsoft.com/en-us/windows/console/coord-str) but you use this in our syscall binding generation as it can't be converted into a uintptr. This just makes a buffer with the required values and casts it into a uintptr